### PR TITLE
Suggest deriving required supertraits

### LINKED ISSUE #91550: Suggestion to derive traits does not consider needing to implement supertraits
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8d2ebff94e06c67049c0ad3f096873ff

```rust
use std::collections::HashSet;

struct Value(u32);

fn main() {
    let hs = HashSet::<Value>::new();
    hs.insert(Value(0));
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: the method `insert` exists for struct `HashSet<Value>`, but its trait bounds were not satisfied
 --> src/main.rs:7:8
  |
3 | struct Value(u32);
  | ------------------
  | |
  | doesn't satisfy `Value: Eq`
  | doesn't satisfy `Value: Hash`
...
7 |     hs.insert(Value(0));
  |        ^^^^^^ method cannot be called on `HashSet<Value>` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `Value: Eq`
          `Value: Hash`
help: consider annotating `Value` with `#[derive(Eq, Hash)]`
  |
3 | #[derive(Eq, Hash)]
  |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: the method `insert` exists for struct `HashSet<Value>`, but its trait bounds were not satisfied
 --> src/main.rs:7:8
  |
3 | struct Value(u32);
  | ------------------
  | |
  | doesn't satisfy `Value: Eq`
  | doesn't satisfy `Value: Hash`
...
7 |     hs.insert(Value(0));
  |        ^^^^^^ method cannot be called on `HashSet<Value>` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `Value: Eq`
          `Value: Hash`
help: consider annotating `Value` with `#[derive(PartialEq, Eq, Hash)]`
  |
3 | #[derive(PartialEq, Eq, Hash)]
  |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```

This way, the suggested code will compile without the user needing to know that in order to derive Eq, you must implement PartialEq.

There's also an issue where you need the `::<Value>` turbofish in the above code for the suggestion to show up, I'll make a separate issue for that.

### PR DESCRIPTION
**Pull request by rukai**
closes https://github.com/rust-lang/rust/issues/91550

I chose to just hardcode handling for PartialOrd and PartialEq because that should be robust enough and I dont know how to go about doing it generically

r? rust-lang/diagnostics

### CODE REVIEW

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 1]
```diff
@@ -1211,10 +1212,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 _ => false,
             };
             if can_derive {
+                let self_name = trait_pred.self_ty().to_string();
+                let self_span = self.tcx.def_span(adt.did);
+                if let Some(sym::Ord) = diagnostic_name {
```
Comment (davidtwco) [inline compiler/rustc_typeck/src/check/method/suggest.rs:1217]: I think we should make an attempt at making this more generic than just `Ord` and `PartialOrd` - could you try [`rustc_middle::traits::util::supertraits`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/traits/util/fn.supertraits.html) to retrieve the supertraits? You might be able to call [`to_opt_poly_trait_ref`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.Predicate.html#method.to_opt_poly_trait_ref) on `pred` to get the `PolyTraitRef` that you need.

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is a good start and I like the diagnostic improvement, I've left a comment with a suggestion on how we might make this more generic, but otherwise happy with this.

File: compiler/rustc_typeck/src/check/method/suggest.rs (hunk 1 above)
Comment (rukai) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: I implemented as you suggested and it caught a whole bunch of cases I missed!

I'm still a bit concerned about some traits having a Sized parent e.g. Clone
I'm not sure at what point the Sized traits get filtered out by the suggestions and if we can rely on that to occur in the future.
I guess thats why we have tests though!
Overall im happy with this approach.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 2]
```diff
@@ -1196,26 +1196,43 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                    self.tcx.def_span(adt.did),
-                    format!("{}", trait_pred.trait_ref.print_only_trait_name()),
-                ));
+                    | sym::Debug => true,
+                    _ => false,
+                };
+                if can_derive {
+                    let self_name = trait_pred.self_ty().to_string();
+                    let self_span = self.tcx.def_span(adt.did);
+                    use crate::rustc_middle::ty::ToPolyTraitRef;
```
Comment (davidtwco) [inline compiler/rustc_typeck/src/check/method/suggest.rs:1215]: nit: you can just put this `use` at the top of the file

File: compiler/rustc_typeck/src/check/method/suggest.rs (hunk 1 above)
Comment (davidtwco) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: I think this is a good improvement and if we end up with some incorrect suggestions related to `Sized` then we can fix that in a follow-up.

**Review by davidtwco — APPROVED** [review]
Looks good to me, if you want to fix the nit then go ahead and I'll approve, otherwise I'll do so in a day or two.

File: compiler/rustc_typeck/src/check/method/suggest.rs (hunk 2 above)
Comment (rukai) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: Oops, missed those, should be good now.

**Pull request merged**